File: CsharpProjects/myLeetCode/Linked_List/19. Remove Nth Node From End of List.cs
// 19. Remove Nth Node From End of List
// https://leetcode.com/problems/remove-nth-node-from-end-of-list/
using System;

public class ListNode {
    public int val;
    public ListNode next;
    public ListNode(int x) { val = x; }
}

public class Solution {
    public ListNode RemoveNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode first = dummy;
        ListNode second = dummy;

        for (int i = 1; i <= n + 1; i++) {
            first = first.next;
        }

        while (first != null) {
            first = first.next;
            second = second.next;
        }
        second.next = second.next.next;

        return dummy.next;
    }
}

public class Program {
    public static void Main() {
        int[] nums = {1, 2, 3, 4, 5};
        ListNode head = ListHelpers.ConvertArrayToList(nums);

        Solution solution = new Solution();
        ListNode result = solution.RemoveNthFromEnd(head, 2);

        Console.Write("Output: ");
        while (result != null) {
            Console.Write(result.val + " ");
            result = result.next;
        }
    }
}

File: LeetCode/19-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# Helper function to print the linked list
def print_linked_list(head):
    current = head
    while current:
        print(current.val, end=" -> " if current.next else "\n")
        current = current.next

# Solution class with the removeNthFromEnd method
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummy = ListNode(0)
        dummy.next = head
        fast = slow = dummy
        
        for _ in range(n):
            fast = fast.next
        
        while fast.next:
            fast = fast.next
            slow = slow.next
        
        slow.next = slow.next.next
        
        return dummy.next

# Create a linked list: 1 -> 2 -> 3 -> 4 -> 5
head = ListNode(1)
head.next = ListNode(2)
head.next.next = ListNode(3)
head.next.next.next = ListNode(4)
head.next.next.next.next = ListNode(5)

print("Original linked list:")
print_linked_list(head)

# Remove the 2nd node from the end (which is node with value 4)
solution = Solution()
new_head = solution.removeNthFromEnd(head, 2)

print("Linked list after removing 2nd node from the end:")
print_linked_list(new_head)
